### DISCUSSION TITLE
Problem with conditional and log point breakpoint - debugging Scala

### DISCUSSION THREAD

**Discussion by pejax** [mfussenegger/nvim-dap#438] [Q&A]
First of all - thanks all for putting effort into this project :-)

I am trying to debug a Scala 2.13.8 (on 17.0.1 JVM) program. Toggling a breakpoint (marked B in the gutter) works fine. When I try to use a conditional breakpoint or a log point the program does nothing visible, although it does not consume process time. All I can do is "continue()" to load the menu and then hit terminate. 
Using:
- Conditional breakpoint (marked C in the gutter): lua require'dap'.set_breakpoint(vim.fn.input('Breakpoint condition: '))
- Log point (marked L in the gutter): lua require'dap'.set_breakpoint(nil, nil, vim.fn.input('Log point message: '))
Any idea where I have to search? Is this maybe a metals problem?

When I open the Dap console, it is empty. I have to type one of the nvim insert characters (a,A,i,I) to get the "dap>" prompt. It seems that the window is not in insert mode. Any idea what's wrong?

Thanks and regards
Peter

**Comment by mfussenegger** [comment]
> Is this maybe a metals problem?

Could be. Some debuggers output the log point messages to the stdout of the application you're debugging. If you're running the application in an integrated terminal or external and are attaching the debugger to it, the output could be visible there.

Otherwise you could check the logs. See `:h dap.set_log_level()`


> When I open the Dap console, it is empty. I have to type one of the nvim insert characters (a,A,i,I) to get the "dap>" prompt. It seems that the window is not in insert mode. Any idea what's wrong?

That's normal that it doesn't start in insert mode

**Comment by pejax** [comment]
I set a log point (L in the gutter) and a conditional bp (C in the gutter) and configured dap.set_log_level('TRACE'). The debuggee does not stop at either breakpoint. This is the tail -f dap.log:

[ DEBUG ] 2022-02-16T01:14:12Z+0100 ] ...nvim/site/pack/packer/start/nvim-dap/lua/dap/session.lua:560 ]	{
  body = {
    breakpoints = { {
        id = 1,
        line = 5,
        message = "",
        source = {
          name = "Main.scala",
          path = "/Users/peter/ramdisk/debug/src/main/scala/example/Main.scala"
        },
        verified = true
      }, {
        id = 2,
        line = 7,
        message = "",
        source = {
          name = "Main.scala",
          path = "/Users/peter/ramdisk/debug/src/main/scala/example/Main.scala"
        },
        verified = true
      } }
  },
  command = "setBreakpoints",
  request_seq = 5,
  seq = 6,
  success = true,
  type = "response"
}

Shouldn't the log point breakpoint have the 'message' field filled and the conditional breakpoint have a valid 'condition' field?

**Reply by mfussenegger** [reply]
The log entry you posted is the response to the `setBreakpoints` request. The request should contain a `condition` for conditional breakpoint or a `logMessage` for log points.

**Comment by pejax** [comment]
It helps if one can read - I get this as the request :-)
[ DEBUG ] 2022-02-16T19:37:22Z+0100 ] ...nvim/site/pack/packer/start/nvim-dap/lua/dap/session.lua:850 ]	"request"	{
  arguments = {
    breakpoints = { {
        line = 5,
        logMessage = "Logging..."
      }, {
        condition = "y=9",
        line = 7
      } },
    lines = { 5, 7 },
    source = {
      name = "Main.scala",
      path = "/Users/peter/ramdisk/scala-debug/src/main/scala/example/Main.scala"
    }
  },
  command = "setBreakpoints",
  seq = 2,
  type = "request"
}

The app does not stop at the breakpoints. When I call "continue" I get a menu (Terminate, Pause, etc...), terminate and run it again. When I set a breakpoint without log point or condition, it does. stop. Any idea what else I can check or is this maybe a metals problem?
Thanks

**Reply by mfussenegger** [reply]
From the log it looks like nvim-dap is sending the correct payload, so I'd assume it is a metals problem. Maybe @ckipp01 knows more.

**Reply by ckipp01** [reply]
Yup, this is on us. Thanks for the report @pejax, and I see the issue you created in `nvim-metals` as well. The short answer is that we set in our capabilities that we support `"supportsConditionalBreakpoints": true`, which is actually incorrect, and it's not supported.

There is actually a long chain here of `metals -> BSP Server -> DAP server` where many times the latter two are the same. Either way, any of the BSP servers that you're using are actually using [`scala-debug-adapter`](https://github.com/scalacenter/scala-debug-adapter) under the hood and the fix will need to be there. I already chatted to them over there so they're aware.  I can include more details in the `nvim-metals` report and we can continue there.

Thanks for the ping @mfussenegger and sorry for the noise! 

BTW the issue created for the incorrect init options is [here](https://github.com/scalacenter/scala-debug-adapter/issues/195).
